src/render.rs

```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
         Ok(output.trim_right().to_string())
     }
 
+    fn render_macro(&mut self, call_node: Node) -> TeraResult<String> {
+        if let MacroCall {namespace, name: macro_name, params: call_params} = call_node {
+            // To find the real namespace we just look at the last one we got.
+            // If we have self, we take the last one otherwise we push it into the vec of namespaces
+            let real_namespace = match namespace.as_ref() {
+                "self" => {
+                    // TODO: handle error if we don't have a namespace
+                    self.macro_namespaces
+                        .last()
+                        .expect("Open an issue with a template sample please (mention `self namespace macro`)!")
+                        .to_string()
+                },
+                _ => {
+                    // TODO: String doesn't have Copy trait, can we avoid double cloning?
+                    self.macro_namespaces.push(namespace.clone());
+                    namespace.clone()
+                }
+            };
+
+            if let Some(Macro {body, params, ..}) = self.macros.get(&real_namespace)
+                .and_then(|m| m.get(&macro_name).cloned()) {
+                // fail early if we don't have the same number of args
+                let expected_params = params.iter().cloned().collect::<Vec<String>>();
+                let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
+                // remove that check if we add kwargs to macros
+                if expected_params.len() != params_seen.len() {
+                    return Err(MacroCallWrongArgs(macro_name, expected_params, params_seen));
+                }
+
+                // We need to make a new context for the macro
+                let mut context = HashMap::new();
+                for (param_name, exp) in call_params {
+                    if !params.contains(&param_name) {
+                        return Err(MacroCallWrongArgs(macro_name, expected_params, params_seen));
+                    }
+                    context.insert(param_name.to_string(), try!(self.eval_expression(exp)));
+                }
+                self.macro_context = Some(to_value(&context));
+                let mut output = String::new();
+                for node in body.get_children() {
+                    output.push_str(&try!(self.render_node(node)));
+                }
+                // Remove the namespace before returning if necessary
```

Comment:
This comment doesn't help me understand why this is necessary or important or correct.